gobblin-service/src/main/java/org/apache/gobblin/service/modules/utils/FlowCompilationValidationHelper.java

```diff
@@ -0,0 +1,196 @@
+ * carry out any flow action requests. However, with launch executions now being stored in the DagActionStateStore, on
+ * restart or leadership change the DagManager has to perform validations before executing any launch actions the
+ * previous leader was unable to complete. Rather than duplicating the code or introducing a circular dependency between
+ * the DagManager and Orchestrator, this class is utilized to store the common functionality. It is stateful,
+ * requiring all stateful pieces to be passed as input from the caller upon instantiating the helper.
+ * Note: We expect further refactoring to be done to the DagManager in later stage of multi-active development, so we do
+ * not attempt major reorganization as abstractions may change.
+ */
+@Slf4j
+@AllArgsConstructor
+public final class FlowCompilationValidationHelper {
+  private SharedFlowMetricsSingleton sharedFlowMetricsSingleton;
+  private SpecCompiler specCompiler;
+  private UserQuotaManager quotaManager;
+  private Optional<EventSubmitter> eventSubmitter;
+  private FlowStatusGenerator flowStatusGenerator;
+  private boolean isFlowConcurrencyEnabled;
+
+  /**
+   * For a given a flowSpec, verifies that an execution is allowed (in case there is an ongoing execution) and the
+   * flowspec can be compiled. If the pre-conditions hold, then a JobExecutionPlan is constructed and returned to the
+   * caller.
+   * @return jobExecutionPlan dag if one can be constructed for the given flowSpec
+   */
+  public Optional<Dag<JobExecutionPlan>> createExecutionPlanIfValid(FlowSpec flowSpec)
+      throws IOException, InterruptedException {
+    Config flowConfig = flowSpec.getConfig();
+    String flowGroup = flowConfig.getString(ConfigurationKeys.FLOW_GROUP_KEY);
+    String flowName = flowConfig.getString(ConfigurationKeys.FLOW_NAME_KEY);
+
+    //Wait for the SpecCompiler to become healthy.
+    specCompiler.awaitHealthy();
+
+    Optional<Dag<JobExecutionPlan>> jobExecutionPlanDagOptional =
+        validateAndHandleConcurrentExecution(flowConfig, flowSpec, flowGroup, flowName);
+    if (!jobExecutionPlanDagOptional.isPresent()) {
+      return Optional.absent();
+    }
+
+    Optional<TimingEvent> flowCompilationTimer =
+        eventSubmitter.transform(submitter -> new TimingEvent(submitter, TimingEvent.FlowTimings.FLOW_COMPILED));
+    Map<String, String> flowMetadata = TimingEventUtils.getFlowMetadata(flowSpec);
+
+    if (jobExecutionPlanDagOptional.get() == null || jobExecutionPlanDagOptional.get().isEmpty()) {
+      populateFlowCompilationFailedEventMessage(eventSubmitter, flowSpec, flowMetadata);
+      return Optional.absent();
+    }
+
+    addFlowExecutionIdIfAbsent(flowMetadata, jobExecutionPlanDagOptional.get());
+    if (flowCompilationTimer.isPresent()) {
+      flowCompilationTimer.get().stop(flowMetadata);
+    }
+    return jobExecutionPlanDagOptional;
```

Comment:
granted, the timer makes it a bit tricky (I have thoughts on reworking if you want to discuss.), but this begins to look to me like:
```
import org.apache.commons.lang3.tuple.Pair;
...
return jobDagOptional
    .map(dag -> Pair.of(dag, TimingEventUtils.getFlowMetadata(flowSpec)))
    .filter(p -> {
       Dag dag = p.getLeft()
        if (dag != null && !dag.isEmpty()) {
           return true;
        } else {
           populate(...)
           return false;
       }
    }).map(p -> addFlowExecutionIdIfAbsent(p.getRight(), p.getLeft()));

Reply:
Personally I find this also a bit unintuitive to read especially since we're working with one dag at a time the if/else blocks with the timer are easier to handle in the current state.